File: code/scenes/src/main/java/org/fbme/scenes/controllers/ROLayoutModel.java
package org.fbme.scenes.controllers;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.util.Set;

public interface ROLayoutModel<T> {

    Set<T> components();

    @Nullable
    T findAt(int x, int y);

    @NotNull
    Point getActivePosition(@NotNull T component);

    @Nullable
    Point getTracePosition(@NotNull T component);

    void addListener(@NotNull Listener<T> listener);

    void removeListener(@NotNull Listener<T> listener);

    interface Listener<T> {
        void onComponentAdded(@NotNull T component, int x, int y);

        void onComponentRemoved(@NotNull T component);

        void onComponentsMoved(@NotNull Set<T> components, int dx, int dy, boolean completed);
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/ROLayoutModel.kt
package org.fbme.scenes.controllers

import java.awt.Point

interface ROLayoutModel<T> {
    fun components(): Set<T>
    fun findAt(x: Int, y: Int): T?
    fun getActivePosition(component: T): Point
    fun getTracePosition(component: T): Point?
    fun addListener(listener: Listener<T>)
    fun removeListener(listener: Listener<T>)
    interface Listener<T> {
        fun onComponentAdded(component: T, x: Int, y: Int)
        fun onComponentRemoved(component: T)
        fun onComponentsMoved(components: Set<T>, dx: Int, dy: Int, completed: Boolean)
    }
}
